---
title: "Introduction of the mdist package"
author: "Wolfgang Kopp"
date: "`r Sys.Date()`"
vignette: >
    %\VignetteIndexEntry{Introduction of the mdist package}
    %\VignetteEngine{knitr::rmarkdown}
    \usepackage[utf8]{inputenc}
output:
    BiocStyle::html_document:
    toc: true
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

# Introduction to the mdist package

## Biological background

Transcription factors (TFs) play a crucial role in gene regulation.
They function by recognising and binding to specific DNA stretches
that are usually 5-30bp in length 
and are commonly referred to as *transcription factor binding sites* (TFBSs).
TF-binding acts on the neighbouring
genes by up- or downregulating their gene expression levels.

The aim of the `mdist` package is to provide a sophisticated statistical
tools in order to determine whether a given DNA sequence is enriched
for putative binding sites of a given transcription factor motif.

## Overview of the `mdist`

In order to run an enrichment test using the `mdist` package
three types of user input are required:

    1. a position frequency matrix (PFM)
    2. a DNA sequence for fitting an order-d Markov model as a background model.
    3. a desired false positive probability for motif hits 
        to occur in random DNA sequences.

The statistics computed in `mdist` is based on the log-likelihood 
ratio, also referred to as **score**, between the PFM and the background model.
To this end, `mdist` determines the **score distribution** that would result
from sequence that are generated by the background model using an efficient
dynamic programming algorithm. Importantly, supports the use of higher-order
background models as opposed to only accounting for mono-nucleotide frequencies.
Once the score distribution was determined, the false positive
probability for obtaining a motif hit is used to derive the **score threshold**.

The score threshold is then used to determine the putative TFBSs or **motif
hits** by scanning a DNA sequence, computing the score at each position and
calling a motif hit whenever the score equals or exceeds the score threshold.

In addition to the score distribution, the second important distribution
that is determined in `mdist` is 
**the distribution of the number of motif hits** that would be observed in
random DNA sequences of a given length.
`mdist` provides two efficient algorithms for analytically 
approximating the distribution of the number of motif hits 
in random sequences. The first approach is based on the **compound Poisson 
approximation** and the second is deterined using a novel **combinatorial
model**.
These approximative distributions of the number of motif hits 
are ultimately used as a null model for the motif hit enrichment test. 
Comparing the observed number of motif hits in a DNA sequence of interest
to the null distribution of the number of motif hits is then used
to derive the associated P-value for the motif hit enrichment.

## Hallmarks of the mdist package

In the following we highlight some of the characteristics of `mdist`.
The aim of `mdist` is similar to the `PWMEnrich` bioconductor 
package  [@pwmenrich]. However, while the `PWMEnrich` package 
provides functions for an enrichment test that is based on the binomial
distribution as well as on the log-normal distribution compared to genomic
background, our package follows a different methodological approach.

* A specialty of `mdist` is that it supports the usage of higher-order
    Markov models as background models. To this end, the user is asked to
    provide a set of sequences that are used to fit the background
    and the desired order for the model.
    Higher-order background models
    have also been advocated in other tools, including RSAT [@rsat1], as
    they reflect the structure of real DNA sequences 
    more appropriate  compared to order-0 or CG-balanced background models.
    CpG dinucleotides
    which play an important role in regulatory regions).

* The motif hit enrichment test provided by `mdist` accounts
    for the self-overlapping structure of a given motif. 
    This is 
    is important, because self-overlapping motifs tend to produce
    so-called *clumps* of motif hits. That is, a motif hit gives rise
    to multiple mutually overlapping motif hits.
    *motif clumping* affects 
    the distribution of the number of motif hits and consequently
    the motif hit enrichment test.
    Typical self-overlapping
    motifs are palindromic motifs and repeat-like motifs.

* The `mdist` package does not only account for the clumping of motif hits
    on a single DNA strand, but also for clumps of motif hits that 
    arise by scanning both DNA strands. This is useful in cases
    when TFBSs might be located on either case, in which case
    both strands need to be scanned for motif hits.

* Most of the algorithms in `mdist` were developed in C, in order
    to achieve maximal efficiency.
     
* `mdist` optionally supports openMP and allows the user to choose the
    desired number of threads  that should be allocated to runtime critical
    computations.


# Typical workflow for an enrichment analysis

## Preliminary settings

Prior to the enrichment analysis, the user is required to 
define some `mdist`-specific options by running

```r
alpha=0.001   # false positive probability of observing a motif hit
gran=0.1      # Score granularity
ncores=1      # Number of cores to use if openMP is available
mdistOption(alpha, gran, ncores)
```
`alpha` is used to set the desired false positive probability for motif hits
in random sequences. That is sequences that are generated by 
the background model. The choice `alpha=0.001` would result in 
one false positive motif hit per strand in a sequence of length 1000.
`gran` is used to discretise the real-valued score range. Small choices
of `gran` increase the number of discrete bins to represent the scores, 
which might increase the accuracy of  the computation, but also increases
the runtime of the required algorithms in `mdist`.
The 'ncores' argument defines the number cores that shall be invoked
for parallel computation. This parameter is only relevant if openMP is
supported by the environment.

## Load a TF motif

Next, the user is required to load a PFM into `mdist`.
This is done by invoking `readMotif()` as in the following example

```r
# tab file that contains a PFM
motiffile=system.file("extdata","x31.tab", package="mdist")
readMotif(motiffile)

# transfac file that contains a PFM
pwmfile=system.file("extdata","test.transfac", package="mdist")
readMotif(motiffile)

# R matrix representing a PFM
pfm=motif2matrix()
readMotif(pfm)
```

`readMotif()` requires either a filename for a file that
contains a PFM
in 'transfac' or 'tab' format. In either case, only one PFM
should be contained in the file.
Alternatively, the PFM can also be loaded from an R matrix.

By default, `readMotif` adds a small pseudo-count to each entry
in the PFM and re-normalises the matrix. This is necessary to prevent
entries from being equal to zero. 

## Estimating the background model

In order to estimate the background model, the user needs
to provide a fasta-file and the desired order of the background model.

```r
# Example fasta-file 
seqfile=system.file("extdata","seq.fasta", package="mdist")

readBackground(seqfile,0) # fit an order-0 background model
readBackground(seqfile,1) # fit an order-1 background model
readBackground(seqfile,2) # fit an order-1 background model

```

Higher-order background models might reflect relevant
features of the DNA sequence better, however, this also
increases computational burden for algorithms that are used in `mdist`.

## Perform the motif hit enrichment test

The simplest method to determine motif hit enrichment is by
invoking the high-level function `motifEnrichmentTest`.
In this case, assuming the PFM and the background model have been
loaded already as shown above, one can determine the
motif hit enrichment test in several alternative ways

```r

# Fasta-sequence which shall be assessed for motif enrichment
seqfile=system.file("extdata","seq.fasta", package="mdist")

# Enrichment test based on the compound Poisson distribution and
# assuming both DNA strands are scanned for motif hits
motifEnrichmentTest(seqfile, singlestranded=FALSE, method="compound")

# Enrichment test based on the compound Poisson distribution and
# assuming a single DNA strands is scanned for motif hits
motifEnrichmentTest(seqfile, singlestranded=TRUE, method="compound")

# Enrichment test based on the combinatorial model and
# assuming a single DNA strands is scanned for motif hits
motifEnrichmentTest(seqfile, singlestranded=FALSE, method="combinatorial")
```

The 'combinatorial' and the 'compound' Poisson model are two alternative
ways for approximating the distribution of the number of motif hits.
Each comes with advantages and disadvantages. We recommend to use
the 'compound' option for the following situations:

1. A single or a both DNA strands is scanned for enrichment.
2. A set of variable-length sequences should be scanned for enrichment.
3. The false positive probability of obtaining motif hits should be less than
    1%. For relaxed significance levels, the compound Poisson model
    might yield conservative p-value estimates, because the assumption
    of 'rare motif hit occurrences' is not met any longer.

On the other hand, the 'combinatorial' model can be used in most situations

1. The combinatorial model work particularly well for relaxed 
    choices of the false positive probability for obtaining a motif hit,
    e.g. 5%.
2. At the moment, the 'combinatorial' model is only supported for scanning
    both DNA strands.
3. The set of DNA sequences needs to contain sequences of equal length


