---
title: "Introduction to the `motifcounter` package"
author: "Wolfgang Kopp"
date: "`r Sys.Date()`"
output:
    prettydoc::html_pretty:
        toc: TRUE
        number_sections: TRUE
        fig_caption: yes
        fig_height: 4
        fig_width: 6
vignette: >
    %\VignetteIndexEntry{Introduction to the `motifcounter` package}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
bibliography: literature.bib
---


```{r include=FALSE}
library(knitr)
knit_hooks$set(plot = function(x, options) {
    paste('<figure><img src="',
        opts_knit$get('base.url'), paste(x, collapse = '.'),
        '"><figcaption>', options$fig.cap, '</figcaption></figure>',
        sep = '')
})
library(motifcounter)
library(MotifDb)
library(seqLogo)
opts_chunk$set(fig.path="fig/")
```


This software package grew out of the work that I did to obtain my PhD.
If it is of help for your analysis, please cite
```
@Manual{,
    title = {motifcounter: R package for analysing TFBSs in DNA sequences},
    author = {Wolfgang Kopp},
    year = {2016},
    note = {R package version 0.99.0},
}
```

# Introduction


## Biological background

Transcription factors (TFs) play a crucial role in gene regulation.
They function by recognising and binding to specific DNA stretches
that are usually 5-30bp in length
which are referred to as *transcription factor binding sites* (TFBSs).
TF-binding acts on the neighbouring
genes by up- or downregulating their gene expression levels.

The aim of the `motifcounter` package is to provide statistical
tools for studying putative TFBSs in given DNA sequence, including
the presence and location of TFBSs and the enrichment of TFBSs.

## Hallmarks of `motifcounter`

The main ingredients for an analysis with `motifcounter` consist of

1. a position frequency matrix (PFM) (also called TF motif)
2. a background model
3. a set of DNA sequences that should be scanned for TFBSs
4. a *false positive probability* $\alpha$ to obtain a motif hit 
    in a random sequence

A **PFM** represents the affinity of a TF to bind a certain DNA segment. A
large set of known PFMs can be acquired e.g. from 
the `MotifDb` package [@motifdb].
On the other hand,
the **background model** defines the properties of random DNA sequences.
`motifcounter` implements the background model as an
**order-$d$ Markov model**, where the order $d$ is chosen by the user.
The advantage of using higher-order background models is that they
are able to capture higher-order sequence features which is crucial
for studying real DNA sequences that are rich in
higher-order sequence features (e.g. in CpGs islands).

Using the PFM and the background model, `motifcounter` computes
the **motif score** for a give DNA sequence,
which is defined to be the log-likelihood ratio 
between the PFM and the background.
The motif score represents a measure that indicates whether a certain
position in the DNA sequence is bound or unbound by the TF. Intuitively, 
the higher the score, the more like the location is a TFBSs.

The motif scores are also used to determine
**motif hits** (e.g. putative TFBSs) in the DNA sequence.
To this end, `motifcounter`
uses a pre-determined **score threshold** upon which it is decided
whether a position in the DNA sequence represents a putative TFBS or not.
`motifcounter` establishes the **score threshold**
automatically based on 1) the
**score distribution** and 2) $\alpha$ (the
probability to observe a motif hit in a random DNA sequence).
In `motifcounter`, the score distribution is determined by
an efficient dynamic programming algorithm for general order-$d$
background models, similar as described in [@rahmann, @rsat1].

Testing for **motif hit enrichment** in `motifcounter` is based
on the **number of motif hits** that are observed in a set of DNA sequences.
Consequently, the statistical significance for 
the notif enrichment is based on the
**distribution of the number of motif hits** 
in a set of random DNA sequences.

`motifcounter` provides two fast and accurate analytical
approximations of the distribution of the number of motif hits:

1. A compound Poisson approximation
2. a combinatorial model


The main advantage of those two models (as opposed to using a
simple binomial approximation [@rsat1,@rahmann])
is that they account for the coupling of motif hits due to
mutually **overlapping motif hits**, which are referred to as
**clumps** [@reinert].
**Clumping** is particularly  characteristic for self-overlapping motifs,
including repeat-like motifs or palindromic motifs. It is necessary
to account for clumping, as it influences the distribution
of the number of motif hits and thereby the motif hit enrichment test.

The compound Poisson model in `motifcounter` is based
on an improvement upon [@pape], which was extended to support order-$d$
background models and uses a refined way to 
determine the clump size distribution.

On the other hand,
the combinatorial model is based on summing over all permutations of
placing the $k$ hits in a finite-length sequence, which computed
based on a dynamic programming strategy and is related
to the approach proposed by [@zhang]^[`motifcounter`, however, 
does not seek to enumerate all compatible words that give rise to
a motif hit, but is rather based on a fast approximation
of overlapping hit probabilities. This makes the combinatorial approximation
applicable for arbitrary lengths and structures of motifs as well as rather
relaxed choices of $\alpha$].

`motifcounter` not only accounts for clumps of motif hits
on a single DNA strand, but
also for clumps that result from scanning both DNA strands for motif hits.


# Getting started
## Preliminary steps
### Acquire a background model
The background model is used to specify the 
theoretical properties of random DNA sequences.
`motifcounter` uses the background model to 
derive the theoretical *score distribution*
and the *distribution of the number of motif hits* 
for a given motif (and $\alpha$) in a random DNA sequence.

In `motifcounter`, the background model is implemented 
as an order-$d$ Markov model. A background model is obtained 
from a given set of DNA sequences. Ideally, the DNA sequence for 
estimating the background model should be representative (or even the same)
as the sequences that are latter analysed (e.g. for motif hit enrichment).

Furthermore, `motifcounter` allows the user to choose the order $d$ of the 
background model that should be used for the analysis.
While, higher-order models tend to be more accurate for representing the real
DNA sequence, they come at a higher computational cost. 
Therefore, as a compromise, we recommend to use orders $d=1$ or $d=2$. 


The following code fragment examplifies how an order-$1$ 
background model is estimated from a given set of DNA sequences. 
To this end, the set of DNA sequences must be supplied as
an `DNAStringSet` object from the `Biostrings` bioconductor package.
```{r}
order=1
file=system.file("extdata","seq.fasta", package="motifcounter")
seqs=Biostrings::readDNAStringSet(file)
bg=readBackground(seqs,order)
```

### Acquire a motif
`motifcounter` handles motifs in terms of 
*position frequency matrices* (PFMs), which are commonly
used to represent the binding affinity of transcription factors (TFs).

A convenient source of known motifs is the `MotifDb` 
bioconductor package [@motifdb],
which shall be the basis for our tutorial.
For example, we retrieve the motif for the human *Pou5f1* (or *Oct4*) 
transcription factor as follows


```{r}
library(MotifDb)
oct4=as.list(query(query(query(MotifDb,
            "hsapiens"),"pou5f1"),"jolma2013"))[[1]]
motif=oct4
```

**NOTE:** `motifcounter` requires the PFMs be be matrices 
with strictly positive entries.
If this is not the case, the package provides the 
function `normalizeMotif`, which adds
pseudo-observations to each element of the matrix and renormalize the columns:

```{r}
new_motif=normalizeMotif(motif)
```


### Optional settings

By default the false positive probability $\alpha$ to obtain
a motif hit in random sequences is set to $\alpha=0.001$. The user might change
the stringency level of $\alpha$ globally using the `motifcounterOption`.

```{r eval=FALSE}
alpha=0.01
motifcounterOption(alpha)
```

Please consult `?motifcounterOption` for more information.



## Retrieve position- and strand-specific scores and hits

For the following example, we explore the DNA sequences of 
a set of *Oct4*-ChIP-seq peaks that were obtained in human *hESC* by 
the ENCODE project [@encode2012]. The peak regions were trimmed to 200 bps.

```{r}
file=system.file("extdata","oct4_chipseq.fa", package="motifcounter")
oct4peaks=Biostrings::readDNAStringSet(file)
```

### Analysis of individual DNA sequences

The `motifcounter` package provides functions for exploring 
position- and strand-specific putative TFBSs in individual DNA sequences.
One way to explore the a given DNA sequence for TFBSs is by 
utilizing `scoreSequence`. This function returns the scores at each 
position and strand for a given `Biostring::DNAString`-object (left panel below).
To put the observed scores into perspective, the right panel shows
the theoretical score distribution in random sequences, which
is obtained by `scoreDist`. Scores at the tail of the distribution 
occur very rarely by chance:

```{r fig.show=TRUE, fig.cap="Per-position and strand scores"}
seq=oct4peaks[[1]]
# Determine the per-position and per-strand scores
scores=scoreSequence(seq,motif,bg)
# As a comparison, compute the theoretical score distribution
sd=scoreDist(motif,bg)
par(mfrow=c(1,2))
plot(1:length(scores$fscores),scores$fscores, type="l",
    col="blue",xlab="position",ylab="score",
    ylim=c(min(sd$score),max(sd$score)),xlim=c(1,250))
points(scores$rscores,col="red",type="l")
legend(120,11,c("forw.", "rev."), col=c("blue","red"),lty=c(1,1))

# plot the the score distribution for the comparison
plot(sd$probability,sd$score, type="l",xlab="probability",ylab="")
```

Alternatively, `motifcounter` provides an analogous function for determining
the per-position and per-strand motif hit occurrences 
using `motifHits` that emerge with a stringency level $\alpha=0.001$ 
(`motifcounter`'s default setting).

```{r}
seq=oct4peaks[[1]]
mhits=motifHits(seq,motif,bg)
fhitpos=which(mhits$fhits==1)
rhitpos=which(mhits$rhits==1)
fhitpos
rhitpos
```
In the given sequence, we obtain no motif hit on the forward strand and 
one motif hit on the reverse strand at position 94. 
The underlying DNA sequence at this hit can be retrieved by

```{r}
seq[rhitpos:(rhitpos+ncol(motif)-1)]
```

Relaxing stringency for obtaining motif hits
(e.g. to a false positive probability of $\alpha=0.01$ 
via `motifcounterOption`)
will tend to produce more motif hits.

```{r}
motifcounterOption(alpha=0.01)

mhits=motifHits(seq,motif,bg)

fhitpos=which(mhits$fhits==1)
rhitpos=which(mhits$rhits==1)
fhitpos
rhitpos
```
For example, now we observe four hits on either strand 
in the same sequence which was used above.

### Analysis of a set of DNA sequences

While, `scoreSequence` and `motifHits` can be applied to study 
TFBSs in `DNAString`-objects, one might also be interested in the 
average score or motif hit profiles across multiple sequences of equal length.
This might reveal positional constraints of the motif occurrences 
with respect to e.g. the TSS, or the middle or summit of the ChIP-seq peaks.
To study positional constrains, `motifcounter` provides the `scoreSequenceProfile` 
method which can be applied for a `Biostrings::DNAStringSet`-object.

```{r fig.show=TRUE, fig.cap="Average score profile"}

scores=scoreSequenceProfile(oct4peaks,motif,bg)

plot(1:length(scores$fscores),scores$fscores, type="l",
    col="blue",xlab="position",ylab="score")
points(scores$rscores,col="red",type="l")
legend(-10.8,0,c("forward", "reverse"), col=c("blue","red"),lty=c(1,1))
```

Furthermore, `motifHitProfile` constructs a similar profile by computing
the position and strand specific mean motif hit frequency

```{r fig.show=TRUE, fig.cap="Average motif hit profile"}
motifcounterOption() # lets use the default alpha again

mhits=motifHitProfile(oct4peaks,motif,bg)

plot(1:length(mhits$fhits),mhits$fhits, type="l",
    col="blue",xlab="position",ylab="score")
points(mhits$rhits,col="red",type="l")
legend(0,0,c("forward", "reverse"), col=c("blue","red"),lty=c(1,1))
```


## Test for motif hit enrichment

A fundamental task in Bioinformatics concerns the task of determining
statistically enriched TFBS occurrences in DNA sequences.
To this end, `motifcounter` determines a fast and accurate approximation
of the *distribution of the number of motif hits* based on either
the **compound Poisson distribution** or the **combinatorial distribution**.

Motif hit enrichment can be determined using the function `motifEnrichment`,
which requires a set of DNA sequences of interest, a PFM and a background model.

```{r}
#take a subset of the oct4 peaks
seqs=oct4peaks[1:10]

# Enrichment of Oct4 in Oct4-ChIP-seq peaks
result=motifEnrichment(seqs, motif,bg)
result
```

The result of this analysis is a list that contains 
`pvalue` as well as `fold`, where `fold`
represents the fold-enrichment of motif hits with respect to the 
expected number of hits that emerge in random DNA sequences of the
same length. 

**Note:** Be aware that if the expected number of motif hits very high
(e.g. >1000), the *P-values* that are produced will almost always tend 
towards zero, because the distribution of the number of motif hits
becomes very narrow. In this case, the fold-enrichment gives an indication
as to how strongly the DNA is enriched for motif hits, nevertheless.

By default, `motifEnrichment` scans both DNA strands for motif hits and  
draws its statistical conclusions based on the compound Poisson model.
Please consult `?motifEnrichment` if only a single strand should be scanned
for enrichment or if the combinatorial model should be invoked for the
analysis.


### When to use the combinatorial or the compound Poisson model

The combinatorial model and the compound Poisson approximation both aim
towards modelling the distribution of the number of motif hits and
will mostly yield similar results, but they come 
advantages and disadvantages as well.
In this section, we describe in which cases each approximation is expected to
yield accurate results and in which cases the underlying assumptions for each 
method might be violated. In the latter case, the results of the enrichment 
analysis might be misleading.

If there is no strong preference in changing the default parameters,
the compound Poisson approxmiation is most likely your the method of choice.

The compound Poisson approximation is based on the assumption that motif hits
occur rarely in random sequences. This is also referred to as the Poisson 
assumption.  This means that one should use rather stringent 
false positive probabilities $\alpha=0.001$ or smaller.  Relaxed 
stringency levels result in a conservative *P-value* estimates, 
as in this case the approximation tends to over-estimation of 
the number of motif hits.

The main benefit of the combinatorial model is that it does not presume 
a rare number of hits, because of the different algorithmic approach.
Therefore, it can be used for rather relaxed choices 
of $\alpha$ (e.g. even $\alpha=0.05$). However, the combinatorial model 
also serves as an accurate model for fairly stringent choices of $\alpha$, 
in which the results are very similar to the compound Poisson approximation.
The downside of the combinatorial model is, however, the 
comparably high computational burden. Moreover, the combinatorial 
model is restricted to analysing only fixed-length sequences  
and at the moment only supports scanning for motif hits
in both DNA strands.


