#' Overlapping motif hit probabilities
#' 
#' This function computes the self-overlapping probabilites of the motif for
#' the current PFM and background based on two-dimensional score distributions.
#' Subsequently, the overlapping hit probabilities are corrected for motif hits
#' that would arise in between the shifted positions.
#' 
#' 
#' @param singlestranded Boolian which defines whether the overlapping hit
#' probabilities shall be computed with respect to scanning both DNA strands or
#' only one strand.  Default: Both strands are scanned for DNA sequences.
#' @return A list containing various overlapping hit probabilities.
#' The list contains the following entries
#'     \describe{
#'     \item{alpha}{False positive motif hit probability}
#'     \item{beta}{Vector of overlapping hit probability for hits 
#'             occurring on the same strand. Each element corresponds to
#'             relative distance of the starting positions between the
#'             motif hits.}
#'     \item{beta3p}{Vector of overlapping hit probability for a 
#'             forward strand hit that is followed by a reverse strand
#'             hit. Each element corresponds to
#'             relative distance of the starting positions between the
#'             motif hits.}
#'     \item{beta5p}{Vector of overlapping hit probability for a
#'             reverse strand hit that is followed by a forward strand
#'             hit. Each element corresponds to
#'             relative distance of the starting positions between the
#'             motif hits.}
#'     \item{gamma}{Vector of overlapping hit probabilities across
#'             all configurations. In contrast to beta, beta3p and beta5p,
#'             gamma is not corrected for intermediate motif hit events.
#'             only used for the compound Poisson variant according to Pape
#'             et al. 2008.}
#'     \item{singlestranded}{returns to singlestranded flag}
#'     }
#'              
#' @examples
#' 
#' 
#' seqfile=system.file("extdata","seq.fasta", package="mdist")
#' motiffile=system.file("extdata","x31.tab", package="mdist")
#' alpha=0.001
#' gran=0.1
#' mdistOption(alpha, gran)
#' 
#' # estimate background model from seqfile
#' readBackground(seqfile,1)
#' 
#' # load motif model from motiffile
#' readMotif(motiffile,0.01)
#' 
#' # compute the overlap probabilities for scanning both DNA strands
#' op=probOverlapHit(singlestranded=FALSE)
#' 
#' # compute the overlap probabilities for scanning a single DNA strand
#' op=probOverlapHit(singlestranded=TRUE)
#' 
#' @export
probOverlapHit=function(singlestranded=FALSE) {
    mlen=motifLength()
    alpha=numeric(1)
    beta=numeric(mlen)
    beta3p=numeric(mlen)
    beta5p=numeric(mlen)
    gamma=numeric(3*mlen)
    if (singlestranded==TRUE) {
        res=.C("mdist_overlapSingleStranded", alpha, 
            beta, beta3p, beta5p, gamma,PACKAGE="mdist")
    } else {
        res=.C("mdist_overlap", alpha, 
            beta, beta3p, beta5p, gamma,PACKAGE="mdist")
    }
    return (list(alpha=res[[1]],beta=res[[2]],beta3p=res[[3]],beta5p=res[[4]],
        gamma=res[[5]], singlestranded=singlestranded))
}



#' Compound Poisson Approximation
#' 
#' This function computes the distribution of the number of motif hits 
#' that would be observe in a theoretical sequence of length 'L'
#' that is generated by the background model.
#' The distribution can be determine in two alternative ways: 
#' 1) We provide a reimplemented version of the algorithm that was
#' described in Pape et al. \emph{Compound poisson approximation 
#' of the number of occurrences of a position 
#' frequency matrix (PFM) on both strands.} 2008. In contrast to the
#' original model, this implementation can be used with higher-order
#' background models. This version
#' is used with method='pape'.
#' 2) We provide an improved compound Poisson approximation that
#' uses more appropriate statistical assumptions concerning 
#' overlapping motif hits and that can be used with higher-order
#' background models as well. The improved version can be used
#' with method='kopp'.
#' 
#' 
#' @param seqlen Vector of integers that contains the individual sequence
#' lengths of a given set of sequences. This information can be extracted for 
#' given DNA sequence of interest using \code{\link{numMotifHits}}.
#' In contrast to \code{\link{combinatorialDist}}, this function
#' supports variable length sequence to be considered.
#' @param overlap Overlapping hit probabilities. This argument is the result of
#' \code{link{probOverlapHit}}, which is essential for both.
#' @param method String that defines which method shall be invoked 'pape' or
#' 'kopp' (see description). Default: method='kopp'
#'
#' @return List containing
#' \describe{
#' \item{dist}{Distribution of the number of hits}
#' }
#' @examples
#' 
#' 
#' seqfile=system.file("extdata","seq.fasta", package="mdist")
#' motiffile=system.file("extdata","x31.tab", package="mdist")
#' alpha=0.001
#' gran=0.1
#' mdistOption(alpha, gran)
#' 
#' # estimate background model from seqfile
#' readBackground(seqfile,1)
#' 
#' # load motif model from motiffile
#' readMotif(motiffile, 0.01)
#' 
#' # compute the distribution for scanning a single DNA strand
#' 
#' #Compute overlapping probabilities
#' op=probOverlapHit(singlestranded=TRUE)
#' 
#' # Computes the distribution of the number of motif hits
#' seqlen=rep(150,100)
#' dist=compoundPoissonDist(seqlen, op)
#' #plot(1:length(dist$dist)-1, dist$dist)
#' 
#' # compute the distribution for scanning both DNA strands
#' 
#' #Compute overlapping probabilities
#' op=probOverlapHit(singlestranded=FALSE)
#' 
#' # Computes the distribution of the number of motif hits
#' seqlen=rep(150,100)
#' dist=compoundPoissonDist(seqlen, op)
#' #plot(1:length(dist$dist)-1, dist$dist)
#' 
#' @seealso \code{\link{combinatorialDist}}
#' @export
compoundPoissonDist=function(seqlen, overlap, method="kopp") {
    # for all practical cases, a maximal clump size of 60
    # should be enough
    maxclumpsize=60
    # determine the max number of hits automatically from the
    # given sequence length.
    # even though it is a little bit of computational overhead,
    # set the maxhits to the total sequence length
    maxhits=sum(seqlen)
    dist=numeric(maxhits+1)
    if (method=="kopp") {
        res=.C("mdist_compoundPoisson_useBeta", overlap$alpha,
            overlap$beta, overlap$beta3p, overlap$beta5p,
            as.numeric(dist), as.integer(length(seqlen)),
            as.integer(seqlen),
            as.integer(maxhits), as.integer(maxclumpsize),
            as.integer(overlap$singlestranded),PACKAGE="mdist")
        dist=res[[5]]
    } else if (method=="pape") {
        if (overlap$singlestranded==TRUE) {
            stop("The Pape et al. implementation of the compound 
                Poisson distribution can only be for scanning both DNA strands.
                Use probOverlapHit(singlestranded=F).")
        }
        res=.C("mdist_compoundPoissonPape_useGamma", overlap$gamma,
            as.numeric(dist), as.integer(length(seqlen)), as.integer(seqlen),
            as.integer(maxhits), as.integer(maxclumpsize),PACKAGE="mdist")
        dist=res[[2]]
    } else {
        stop("The method must be 'kopp' or 'pape'")
    }
    return (list(dist=dist))
}

